Foundation/Process.swift

```diff
@@ -978,4 +979,133 @@ private func posix(_ code: Int32) {
     default: fatalError("POSIX command failed with error: \(code)")
     }
 }
+
+#if os(Android)
+// Android doesn't provide posix_spawn APIs until recent API level, so we cannot
+// depend on them, but we can imitate the API, and perform the same work.
+
+private struct posix_spawn_file_actions_t {
+    enum Action {
+    case dup2(old: Int32, new: Int32)
+    case close(fd: Int32)
+    }
+
+    private(set) var actions: [Action] = []
+
+    mutating func add(_ action: Action) {
+        actions.append(action)
+    }
+}
+
+private struct posix_spawnattr_t {}
+
+private func posix_spawn_file_actions_init(_ actions: inout posix_spawn_file_actions_t) -> Int32 {
+    // NOTE: Nothing to do here, the struct initializes itself.
+    return 0
+}
+
+@discardableResult
+private func posix_spawn_file_actions_destroy(_ actions: inout posix_spawn_file_actions_t) -> Int32 {
+    // NOTE: Nothing to do here, the struct handles its own memory.
+    return 0
+}
+
+private func posix_spawn_file_actions_adddup2(_ actions: inout posix_spawn_file_actions_t, _ old: Int32, _ new: Int32) -> Int32 {
+    actions.add(.dup2(old: old, new: new))
+    return 0
+}
+
+private func posix_spawn_file_actions_addclose(_ actions: inout posix_spawn_file_actions_t, _ fd: Int32) -> Int32 {
+    actions.add(.close(fd: fd))
+    return 0
+}
+
+private func posix_spawn(_ pidPtr: UnsafeMutablePointer<pid_t>?,
+                         _ launchPath: UnsafePointer<Int8>?,
+                         _ actions: UnsafeMutablePointer<posix_spawn_file_actions_t>?,
+                         _ attrp: UnsafeMutablePointer<posix_spawnattr_t>?,
+                         _ argv: UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>?,
+                         _ envp: UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>?) -> Int32 {
+    // TODO: We completely ignore attrp, because at the moment, the only
+    // invocation doesn't pass a value.
+
+    // Block signals during this fork/execv dance.
+    var signalSet = sigset_t()
+    sigfillset(&signalSet)
+    var oldmask = sigset_t()
+    guard sigprocmask(SIG_BLOCK, &signalSet, &oldmask) == 0 else {
+        fatalError("sigprocmask() call failed: \(errno)")
+    }
+
+    let pid = fork()
+    guard pid == 0  else {
+        // This is the parent. fork call might have been successful or not.
+
+        // Unblock signals.
+        guard sigprocmask(SIG_SETMASK, &oldmask, nil) == 0 else {
+            fatalError("sigprocmask() call failed: \(errno)")
+        }
+
+        if pid < 0 {
+            return pid
+        } else {
+            if let pidPtr = pidPtr {
+                pidPtr.pointee = pid
+            }
+            return 0
+        }
+    }
+
+    // This is the child.
+
+    // Clean up the parent signal handlers
+    for idx in 1..<(NSIG as Int32) {
+        // Seems that SIGKILL/SIGSTOP are sometimes silently ignored, and
```

Comment:
Odd, is that a documented issue on android?

Reply:
Not that I know of. I found it doing trial and error. Some device was fine, some other was crashing here.

In https://pubs.opengroup.org/onlinepubs/009695399/functions/sigaction.html there's discussion that they should be ignored silently, but it also talks about SysV and BSD returning EINVAL. I think the problem is that, in the code I found for Bionic, it seems to delegate to Linux `rt_sigaction`, which is documented to fail for those two signals.

In any case, since `SIGKILL` and `SIGSTOP` cannot be ignored or caught, they are already set to `SIG_DFL`, which is what this code will try to do. Skipping them should not have ill effects.